File: src/place.js
"use strict";
// Place a vendored SPDX license into a project. Copies the canonical SPDX text
// VERBATIM (downloaded/vendored, never authored here) to a destination file
// (default ./LICENSE).
const fs = require("fs");
const path = require("path");
const crypto = require("crypto");
const { getText, meta } = require("./catalogue");

function placeLicense(spdxId, destPath, { force = false } = {}) {
  const m = meta(spdxId);
  if (!m) throw new Error(`Unknown SPDX license id: ${spdxId}`);
  const text = getText(m.licenseId);
  if (text == null) throw new Error(`No vendored text for ${m.licenseId}`);
  const dest = destPath || path.join(process.cwd(), "LICENSE");
  if (fs.existsSync(dest) && !force) {
    throw new Error(`${dest} already exists — pass { force: true } to overwrite`);
  }
  fs.mkdirSync(path.dirname(dest), { recursive: true });
  fs.writeFileSync(dest, text);
  return {
    id: m.licenseId,
    name: m.name,
    dest,
    bytes: Buffer.byteLength(text),
    sha256: crypto.createHash("sha256").update(text).digest("hex"),
  };
}

module.exports = { placeLicense };

File: rust/src/place.rs
//! Place a vendored SPDX license into a project (copies canonical text;
//! never authored here).
use std::fs;
use std::path::Path;

/// Write the canonical SPDX text for `spdx_id` to `dest`. Returns the sha256 of
/// the placed text. Refuses to overwrite unless `force`.
pub fn place_license(spdx_id: &str, dest: &Path, force: bool) -> Result<String, String> {
    let text = crate::catalogue::get_text(spdx_id)
        .ok_or_else(|| format!("Unknown or unavailable SPDX license: {spdx_id}"))?;
    if dest.exists() && !force {
        return Err(format!("{} already exists (pass force to overwrite)", dest.display()));
    }
    if let Some(parent) = dest.parent() {
        let _ = fs::create_dir_all(parent);
    }
    fs::write(dest, &text).map_err(|e| e.to_string())?;
    Ok(crate::notarize::sha256_hex(text.as_bytes()))
}
